src/common/stats/StatsManager.cpp

```diff
@@ -309,6 +309,238 @@ StatusOr<StatsManager::VT> StatsManager::readHisto(const std::string& counterNam
     return sm.histograms_[index].second->getPercentileEstimate(pct, level);
 }
 
+
+// static
+constexpr bool StatsManager::isStatIndex(int32_t i) {
+    return i > 0;
+}
+
+// static
+constexpr bool StatsManager::isHistoIndex(int32_t i) {
+    return i < 0;
+}
+
+// static
+constexpr std::size_t StatsManager::physicalStatIndex(int32_t i) {
+    DCHECK(isStatIndex(i));
+    return i - 1;
+}
+
+// static
+constexpr std::size_t StatsManager::physicalHistoIndex(int32_t i) {
+    DCHECK(isHistoIndex(i));
+    return -(i + 1);
+}
+
+// static
+StatusOr<StatsManager::ParsedName>
+StatsManager::parseMetricName(folly::StringPiece metricName) {
+    std::string name;
+    StatsMethod method;
+    TimeRange range;
+    std::vector<std::string> parts;
+    folly::split(".", metricName, parts, true);
+    if (parts.size() != 3) {
+        LOG(ERROR) << "\"" << metricName << "\" is not a valid metric name";
+        return Status::Error("\"%s\" is not a valid metric name", metricName.data());
+    }
+
+    name = parts[0];
+
+    if (parts[2] == "60") {
+        range = TimeRange::ONE_MINUTE;
+    } else if (parts[2] == "600") {
+        range = TimeRange::TEN_MINUTES;
+    } else if (parts[2] == "3600") {
+        range = TimeRange::ONE_HOUR;
+    } else {
+        // Unsupported time range
+        LOG(ERROR) << "Unsupported time range \"" << parts[2] << "\"";
+        return Status::Error(folly::stringPrintf("Unsupported time range \"%s\"",
+                                                 parts[2].c_str()));
+    }
+
+    folly::toLowerAscii(parts[1]);
+    if (parts[1] == "sum") {
+        method = StatsMethod::SUM;
+    } else if (parts[1] == "count") {
+        method = StatsMethod::COUNT;
+    } else if (parts[1] == "avg") {
+        method = StatsMethod::AVG;
+    } else if (parts[1] == "rate") {
+        method = StatsMethod::RATE;
+    // } else if (parts[1][0] == 'p') {   // TODO(shylock)
+    } else {
+        LOG(ERROR) << "Unsupported statistic method \"" << parts[1] << "\"";
+        return Status::Error(folly::stringPrintf("Unsupported statistic method \"%s\"",
+                                                 parts[1].c_str()));
+    }
+
+    return StatsManager::ParsedName {
+        name,
+        method,
+        range,
+    };
+}
+
+
+void PrometheusSerializer::annotate(
+        std::ostream& out, const std::string metricName, const std::string metricType) const {
```

Comment:
the parameter is value type, this `const` cannot avoid memory copy, you only read the variables why not use the reference type `const std::string&`?

Reply:
A mistake, already correct.